src/tech-record/TechRecordService.ts

```diff
@@ -0,0 +1,128 @@
+							processRecordsRes.provisionalRecords[0].createdTimestamp
+						)
+					: this.techRecordRepository.callGetTechRecords(
+							processRecordsRes.provisionalRecords[1].systemNumber,
+							processRecordsRes.provisionalRecords[1].createdTimestamp
+						);
+		} else {
+			await Promise.reject('Tech record Search returned nothing.');
+		}
+	};
+
+	/**
+	 * helper function is used to process records and count provisional and current records
+	 * @param records
+	 */
+	private groupRecordsByStatusCode = (
+		records: ISearchResult[]
+	): {
+		currentRecords: ISearchResult[];
+		provisionalRecords: ISearchResult[];
+		currentCount: number;
+		provisionalCount: number;
+	} => {
+		const currentRecords: ISearchResult[] = [];
+		const provisionalRecords: ISearchResult[] = [];
+		records.forEach((record) => {
+			if (record.techRecord_statusCode === 'current') {
+				currentRecords.push(record);
+			} else if (record.techRecord_statusCode === 'provisional') {
+				provisionalRecords.push(record);
+			}
+		});
+
+		return {
+			currentRecords,
+			provisionalRecords,
+			currentCount: currentRecords.length,
+			provisionalCount: provisionalRecords.length,
+		};
+	};
+
+	/**
+	 * Retrieves the vehicle weight details for Roadworthisness certificates
+	 * @param testResult
+	 */
+	public async getWeightDetails(testResult: any) {
+		const searchRes = await this.techRecordRepository.callSearchTechRecords(testResult.systemNumber);
+		const techRecord = (await this.processGetCurrentProvisionalRecords(searchRes)) as TechRecordType<
+			'hgv' | 'psv' | 'trl'
+		>;
+		if (techRecord) {
+			const weightDetails: IWeightDetails = {
+				dgvw: techRecord.techRecord_grossDesignWeight ?? 0,
+				weight2: 0,
+			};
+			if (testResult.vehicleType === VEHICLE_TYPES.HGV) {
+				weightDetails.weight2 = (techRecord as TechRecordType<'hgv'>).techRecord_trainDesignWeight ?? 0;
+			} else {
+				if ((techRecord.techRecord_noOfAxles ?? -1) > 0) {
+					const initialValue: number = 0;
+					weightDetails.weight2 = (techRecord.techRecord_axles as any).reduce(
+						(accumulator: number, currentValue: { weights_designWeight: number }) =>
+							accumulator + currentValue.weights_designWeight,
+						initialValue
+					);
+				} else {
+					throw new HTTPError(500, 'No axle weights for Roadworthiness test certificates!');
+				}
+			}
+			return weightDetails;
+		} else {
+			console.log('No techRecord found for weight details');
+			throw new HTTPError(500, 'No vehicle found for Roadworthiness test certificate!');
+		}
+	}
+
+	/**
+	 * Retrieves the adrDetails from a techRecord searched by vin
+	 * @param testResult - testResult from which the VIN is used to search a tech-record
+	 */
+	public getAdrDetails = async (testResult: any) => {
+		const searchRes = await this.techRecordRepository.callSearchTechRecords(testResult.systemNumber);
+		return (await this.processGetCurrentProvisionalRecords(searchRes)) as TechRecordType<'hgv' | 'trl'>;
+	};
```

Comment:
I'm not sure we need the type coercion here, `processGetCurrentProvisionalRecords` takes in a generic argument, therefore this can just be

```
	public getAdrDetails = async (testResult: any) => {
		const searchRes = await this.techRecordRepository.callSearchTechRecords(testResult.systemNumber);
		return await this.processGetCurrentProvisionalRecords<'hgv' | 'trl'>(searchRes);
	};
```
Where we pass in the vehicle type as `T`.

We can also remove the generic `any` argument where this is called and let it be inferred i.e.

`const adrDetails = await this.techRecordService.getAdrDetails(testResult);`

Reply:
Yeah good spot, happy for me to include this in a later PR? (for same reasons as above)

Reply:
Functionally this PR looks good, so happy to defer potential clean ups / improvements for another time. Same goes for all of the comments I've made on this PR. 👍